tests/SolrIndexTest.php

```diff
@@ -508,10 +509,105 @@ public function testShowInSearch()
                 })]
             );
 
-        IndexableService::singleton()->clearCache();
+        SearchableService::singleton()->clearCache();
         SearchUpdater::flush_dirty_indexes();
+    }
+
+    /**
+     * Test that canView() check is used to exclude DataObjects from being added to the index
+     *
+     * Note: this code path that really being tested here is SearchUpdateProcessor->prepareIndexes()
+     * This code path is used for 'inlet' filtering on CMS->save()
+     * The results of this will show-up in SolrIndex->_addAs()
+     */
+    public function testCanView()
+    {
+        // allow anonymous users to assess draft-only content to pass canView() check (will auto-reset for next test)
+        Versioned::set_draft_site_secured(false);
+        Versioned::set_reading_mode('Stage.' . Versioned::DRAFT);
 
-        Versioned::set_reading_mode($defaultMode);
+        $serviceMock = $this->getMockBuilder(Solr4Service::class)
+            ->setMethods(['addDocument', 'deleteById'])
+            ->getMock();
+
+        $index = new SolrIndexTest_ShowInSearchIndex();
+        $index->setService($serviceMock);
+        FullTextSearch::force_index_list($index);
+
+        // will get added
+        $pageA = new Page();
+        $pageA->Title = 'Test Page Anyone';
+        $pageA->CanViewType = 'Anyone';
+        $pageA->write();
+
+        // will get filtered out
+        $page = new Page();
+        $page->Title = 'Test Page LoggedInUsers';
+        $page->CanViewType = 'LoggedInUsers';
+        $page->write();
+
+        // will get added
+        $fileA = new File();
+        $fileA->Title = 'Test File Anyone';
+        $fileA->CanViewType = 'Anyone';
+        $fileA->write();
+
+        // will get filtered out
+        $file = new File();
+        $file->Title = 'Test File LoggedInUsers';
+        $file->CanViewType = 'LoggedInUsers';
+        $file->write();
+
+        // will get added
+        $objOneA = new SolrIndexTest_MyDataObjectOne();
+        $objOneA->Title = 'Test MyDataObjectOne true';
+        $objOneA->ShowInSearch = true;
+        $objOneA->CanViewValue = true;
+        $objOneA->write();
+
+        // will get filtered out
+        $objOne = new SolrIndexTest_MyDataObjectOne();
+        $objOne->Title = 'Test MyDataObjectOne false';
+        $objOne->ShowInSearch = true;
+        $objOne->CanViewValue = false;
+        $objOne->write();
+
+        $callback = function (Apache_Solr_Document $doc) use ($pageA, $fileA, $objOneA): bool {
+            $validKeys = [
+                Page::class . $pageA->ID,
+                File::class . $fileA->ID,
+                SolrIndexTest_MyDataObjectOne::class . $objOneA->ID
+            ];
+            return in_array($this->createSolrDocKey($doc), $validKeys);
+        };
+
+        $serviceMock
+            ->expects($this->exactly(3))
+            ->method('addDocument')
+            ->withConsecutive(
```

Comment:
Have you checked that it fails if there's *more* calls? Can't tell from https://phpunit.de/manual/5.0/en/test-doubles.html#test-doubles.mock-objects

Reply:
that's covered via `$this->exactly(3)`